### PR TITLE
refactor: Fix warnings in CtTypeReferenceImpl

### PR DESCRIPTION
**Pull request by slarse**
#4924 

Most of the fixes are just suppressing of "intentional" warnings.

### CODE REVIEW

**Comment by SirYwell** [comment]
An alternative idea for the unchecked casts: Maybe it would make sense to add a method like
```java
@SuppressWarnings("unchecked")
protected final <T> T self() {
	return (T) this;
}
```
to `CtElementImpl` and replace all the unchecked casts by calls to that method? This way, other (unrelated) unchecked casts don't get supressed accidentally. Additionally, we could write a test to make sure this method is only used in return statements (and with a bit more work, we can likely also make sure it is only called in return statements of methods that should return `this`).

WDYT?

**Comment by slarse** [comment]
@SirYwell Hmm, well on the one hand it's nice for the reason you mentioned. On the other hand, it would also hide something very nasty that we shouldn't be doing in the first place. The explicit '@SuppressWarnings` make it a bit more obvious that in that method, we're up to no good.

Perhaps we could make a helper that's slightly more of an eye-sore, just to make it clear that we really shouldn't be doing that thing. Something like this:

```java
/**
 * Make an unchecked cast of the provided object to the type of the expression it is used in.
 *
 * This should primarily be used to coerce <code>this</code> to the expected return type in methods that are generic
 * in the return type.
 * 
 * THIS IS VERY BAD, use only where legacy requires it to avoid having to litter classes with <code>SuppressWarnings("unchecked")</code>
 */
@SuppressWarnings("unchecked")
public static <T> T uncheckedCastToInferredReturnType(Object object) {
    return (T) object;
}
```

Something like that. It's obviously a horrendous helper mothod. Having it as a separate helper method, rather than providing it through inheritance, also makes it a bit less discoverable to future us (or other integrators who lack this context). It's less likely to be used in a spur of the moment "well that looks neat".

WDYT?

**Comment by SirYwell** [comment]
> On the other hand, it would also hide something very nasty that we shouldn't be doing in the first place.

That's very true, and basically the reason why I suggested to introduce checks to only allow the method to be called in the specific places.
Sadly, we won't get rid of whatever dirty solution we decide on without breaking basically everything, so it's probably worth to invest some work into that.

I don't have a strong opinion on naming and where the method is actually declared. My idea behind inheritance was to avoid unintended usage (as we can pass everything to a method that takes an object), but I can also fully understand your reasoning. So, feel free to implement it the way you suggested.

To prevent misuse, I'd still go with some architecture test on top of it (if you don't want to include it here, we can do that separately I guess), and probably even create a list of the specific methods in which it is allowed (or whatever is suitable to prevent *new* methods from being implemented that way, without slipping through a code review) (?).

File: src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java [hunk 1]
```diff
@@ -841,12 +842,12 @@ public boolean isShadow() {
...
 		getFactory().getEnvironment().getModelChangeListener().onObjectUpdate(this, IS_SHADOW, isShadow, this.isShadow);
 		this.isShadow = isShadow;
-		return (E) this;
+		return Unchecked.castToReturnType(this);
 	}
 
 	@Override
 	public CtTypeReference<T> clone() {
-		return (CtTypeReference<T>) super.clone();
+		return Unchecked.castToReturnType(super.clone());
```
Comment (slarse) [inline src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java:850]: @SirYwell This also seems like an appropriate place to use this, wdyt?

**Comment by slarse** [comment]
@SirYwell I've pushed a new suggestion up. How about that (or something like that, feel free to come with suggestions), and then an architecture test to check that `Unchecked.castToReturnType` is only a) used on arguments `this` or `super.clone()`, and b) is not imported statically (so it's clearly used with the `Unchecked` "namespace").

There are any amount of variations to that strategy, this is just what I came up with.

**Comment by SirYwell** [comment]
> @SirYwell I've pushed a new suggestion up. How about that (or something like that, feel free to come with suggestions)

I like that.

> and then an architecture test to check that `Unchecked.castToReturnType` is only a) used on arguments `this` or `super.clone()`, and b) is not imported statically (so it's clearly used with the `Unchecked` "namespace").

That sounds good. I'd additionally only allow it in return statements, I think that works for our use case (?) and is easy to check.

File: src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java (hunk 1 above)
Comment (SirYwell) [reply src/main/java/spoon/support/reflect/reference/CtTypeReferenceImpl.java:850]: Yep, that makes sense aswell.

**Comment by slarse** [comment]
Closing this for now, perhaps I'll revisit it later.

**Pull request closed without merging**